design_doc.md

```diff
@@ -0,0 +1,268 @@
+# XDGGS - Design document
+
+Xarrays extension for DGGS. Technical specifiactions.
+
+## Goals
+
+The goal of the `xddgs` library is to facilitate working with multiple Discrete Global Grid Systems (DGGSs) via a unified, high-level and user-friendly API that is deeply integrated with [Xarray](https://xarray.dev).
+
+Examples of common DGGS features that `xdggs` should provide or facilitate:
+
+- convert a DGGS from/to another grid (e.g., a DGGS, a latitude/longitude rectilinear grid, a raster grid, an unstructured mesh)
+- convert a DGGS from/to vector data (points, lines, polygons, envelopes)
+- convert between different cell id representations of a same DGGS (e.g., uint64 vs. string)
+- select data on a DGGS by cell ids or by geometries (spatial indexing)
+- change DGGS resolution (upgrade or downgrade)
+- operations between similar DGGS (with auto-alignment)
+- re-organize cell ids (e.g., spatial shuffling / partitioning)
+- plotting
+
+Conversion between DGGS and other grids or vector features may requires specific interpolation or regridding methods.
+
+`xdggs` should leverage the current recommended Xarray extension mechanisms ([apply_ufunc](https://docs.xarray.dev/en/stable/examples/apply_ufunc_vectorize_1d.html), [accessors](https://docs.xarray.dev/en/stable/internals/extending-xarray.html), [custom indexes](https://docs.xarray.dev/en/stable/internals/how-to-create-custom-index.html)) and possibly the future ones (e.g., variable encoders/decoders) to provide DGGS-specific functionality on top of Xarray core features.
+
+`xdggs` should facitiltate interoperability with other existing Xarray extensions (e.g., [xvec](https://github.com/xarray-contrib/xvec) for vector data or [uxarray](https://github.com/UXARRAY/uxarray) for unstructured grids).
+
+`xdggs` should also leverage the existing implementation of various DGGSs exposed to Python via 3rd-party libraries, here below referred as "backends". Preferrably, those backends would expose DGGS functionality in an efficient way as vectorized functions (e.g., working with NumPy arrays).
+
+`xdggs` should try to follow standards and/or conventions defined for DGGS (see below). However, we may need to depart from them for practical reasons (e.g., common practices in popular DGGS libraries that do not fit well with the proposed standards). Strict adherence to a standard is welcome but shouldn't be enforced by all means.
+
+`xddgs` should also try to support applications in both GIS and Earth-System communities, which may each use DGGS in slightly different ways (see examples below).
+
+When possible, `xddgs` operations should scale to fine resolutions (millions of cells) leveraging Xarray interoperability with Dask. This might not be always possible, though. Some operations (spatial indexing) may be hard to support at scale and it shouldn't be a high development priority.
```

Comment:
there's also the idea that whenever necessary (i.e. if the python implementation turns out to be too slow), slow operations need to be written in a lower-level language

Reply:
Yes, for low-level languages / vectorized bindings I also tried to emphasize that in the "goals" and "backends" sections.

Here this sentence is more about horizontal scaling (I can clarify that). While there are some operations that are trivial to support with Dask, other operations like spatial indexing / partitioning / shuffling / sorting are much harder to implement in a distributed fashion. That is not impossible, though, I think we could for example look at [dask-geopandas](https://github.com/geopandas/dask-geopandas). This is pretty hard work, so I wanted to write that it is probably wiser to first try supporting spatial indexing for small amounts of data and then after see how we can scale it horizontally.